src/snapshot/Snapshot.lua

```diff
@@ -0,0 +1,101 @@
+local ReplicatedStorage = game:GetService("ReplicatedStorage")
+
+local Markers = require(script.Parent.Serialize.Markers)
+local Serialize = require(script.Parent.Serialize)
+local deepEqual = require(script.Parent.Parent.deepEqual)
+local ElementKind = require(script.Parent.Parent.ElementKind)
+
+local SnapshotFolderName = "RoactSnapshots"
+local SnapshotFolder = ReplicatedStorage:FindFirstChild(SnapshotFolderName)
+
+local Snapshot = {}
+local SnapshotMetatable = {
+	__index = Snapshot,
+	__tostring = function(snapshot)
+		return Serialize.snapshotDataToString(snapshot.data)
+	end
+}
+
+function Snapshot.new(identifier, data)
+	local snapshot = {
+		_identifier = identifier,
+		data = data,
+		_existingData = Snapshot._loadExistingData(identifier),
+	}
+
+	setmetatable(snapshot, SnapshotMetatable)
+
+	return snapshot
+end
+
+function Snapshot:match()
+	if self._existingData == nil then
+		self:serialize()
+		self._existingData = self.data
+		return
+	end
+
+	local areEqual, innerMessageTemplate = deepEqual(self.data, self._existingData)
+
+	if areEqual then
+		return
+	end
+
+	local newSnapshot = Snapshot.new(self._identifier .. ".NEW", self.data)
+	newSnapshot:serialize()
+
+	local innerMessage = innerMessageTemplate
+		:gsub("{1}", "new")
+		:gsub("{2}", "existing")
+
+	local message = ("Snapshots do not match.\n%s"):format(innerMessage)
+
+	error(message, 2)
+end
+
+function Snapshot:serialize()
+	local folder = Snapshot.getSnapshotFolder()
+
+	local snapshotSource = Serialize.snapshotDataToString(self.data)
+	local existingData = folder:FindFirstChild(self._identifier)
+
+	if not (existingData and existingData:IsA('StringValue')) then
+		existingData = Instance.new("StringValue")
+		existingData.Name = self._identifier
+		existingData.Parent = folder
+	end
+
+	existingData.Value = snapshotSource
+end
+
+function Snapshot.getSnapshotFolder()
```

Comment:
would be nice to give developers control over this - maybe an option to pass in Roact.matchSnapshot